### ISSUE TITLE
Prisma Migrate: Allow data transformations through code (in addition to SQL)

### ISSUE THREAD

**Issue by chainlink** [prisma/prisma#11810]
## Problem

Currently, it seems possible to use SQL files in migrations. For greater flexibility, it would be nice to be able to provide JS files as well.

## Suggested solution

Allow code to be executed in migrations

**Comment by benediktms** [comment]
This would be super useful. There are scenarios where writing raw SQL to manage data migrations is a huge headache without the benefits of JS/TS.

This is especially critical if you are working on a multi tenancy application and need to move client specific data only. Something that you could only access via something like env variables (maybe you can configure this in SQL, I'm not sure, but it definitely would be time consuming and should be delegated to the node tool-chain). 

Another, probably more common use case would be complex data that requires iteration to process properly: Using a for loop in SQL is not nice and it's error prone. 

I think the implementation that TypeOrm or MikroOrm currently use is very nice, where each generated migration is a TS file that will then get converted to SQL via their query runner. This or a similar API would be a huge benefit for Prisma.

**Comment by JVMartin** [comment]
I agree. Migrations should always be ran via the ORM rather than plain SQL. Anything else limits you SO much, for 0 benefit.

Take this example from a TypeORM migration:
```ts
# database/migrations/AdsAbstraction1634246375736.ts

import { MigrationInterface, QueryRunner } from 'typeorm';

import { WorkflowEntity } from '../../../email-platform/entities/workflow.entity';
import { fixWorkflowUi } from '../../shared/utils/temp';

export class AdsAbstraction1634246375736 implements MigrationInterface {
  public async up(queryRunner: QueryRunner): Promise<void> {
    queryRunner.connection.logger.logMigration(AdsAbstraction1634246375736.name);

    await queryRunner.query(`
      ALTER TABLE "workflow-broadcast"
      RENAME COLUMN "fetchCarInsuranceAds" to "fetchAds"
    `);

    const workflowRepository = queryRunner.connection.getRepository(WorkflowEntity);

    const workflows = await workflowRepository.find();

    for (const workflow of workflows) {
      const workflowDocument = workflow.workflowDocument;
      fixWorkflowUi(workflowDocument);
      await workflowRepository.update({ id: workflow.id }, { workflowDocument });
    }
  }

  public async down(_: QueryRunner): Promise<void> {
    //
  }
}
```

Note that in addition to the plain SQL queries, I can use the ORM itself to loop through individual records and update them with extremely complex TS algorithms. This is absolutely impossible in plain SQL, and it's trival / easy using the ORM.

We should convert Prisma to always use ORM migrations instead of SQL migrations, similar to nearly all other major ORMs in all major languages.

**Comment by bocklighting** [comment]
@JVMartin Agree, having plain sql will make migrations unnecessarily harder, if there's a change in the data during the upgrade it needs to happen at the time of schema changes. not having this functionality within Prisma requires the developer to have a separate method for handling data changes after a schema change.

**Comment by acisser** [comment]
One problem that needs to be solved if planning to make Prisma Client available for data migrations is related to schema versioning. So the schema file should become part of the migration assets together with ts/js and sql files. So that each migration is pinned to a specific compatible version of the Prisma generated client.

**Comment by demedos** [comment]
This would help a lot in backfilling data with cuid values when adding non nullable columns to a table filled with data

**Comment by RDanned** [comment]
Another orms migrations are implemented using orm, so I see no problems to make same in prisma. Look at laravel, typeorm or django/flask(SQL alchemy). Sometimes there is a need to run action before or after migrations, and I don't think that it is good idea to run them in clean SQL.

**Comment by Softjey** [comment]
Hi everyone! 👋

I’ve developed a small library to help address this problem: [Prisma Data Migrations](https://www.npmjs.com/package/prisma-data-migrations?activeTab=readme). This library enhances Prisma’s migration system by enabling the use of post-migration scripts for data migrations alongside schema changes. And allows migrate to specific migration using CLI.

You can find the repository here: [GitHub Repository](https://github.com/Softjey/prisma-data-migrations).

Feel free to open issues, report bugs, or suggest improvements. I’m also open to pull requests if you’d like to contribute to the project!

Looking forward to your feedback! 🚀

**Comment by spersico** [comment]
Hey folks, in the same line as @Softjey (congrats btw!), I made my own approach, but instead of extending Prisma, I decided to build a tool to **migrate away** from Prisma Migrations and just use Knex for that (while keeping Prisma as an ORM). I believe that there are pros and cons for each approach.

Feel free to check it out: [prisma-migration-migrator](https://github.com/spersico/prisma-migration-migrator).

Feedback and issues and contributions are appreciated!
